src/SAWScript/Interpreter.hs

```diff
@@ -4969,18 +5034,28 @@ primitives = Map.fromList
     bicVal f opts bic = toValue (f bic opts)
 
 
-filterAvail ::
+filterAvailTypes ::
+  Set PrimitiveLifecycle ->
+  Map SS.Name Primtype ->
+  Map SS.Name Primtype
+filterAvailTypes primsAvail =
+  Map.filter (\p -> primtypeLife p `Set.member` primsAvail)
+
+filterAvailPrims ::
   Set PrimitiveLifecycle ->
   Map SS.LName Primitive ->
   Map SS.LName Primitive
-filterAvail primsAvail =
+filterAvailPrims primsAvail =
   Map.filter (\p -> primitiveLife p `Set.member` primsAvail)
 
-primTypeEnv :: Set PrimitiveLifecycle -> Map SS.LName SS.Schema
-primTypeEnv primsAvail = fmap primitiveType (filterAvail primsAvail primitives)
+primValueTypeEnv :: Set PrimitiveLifecycle -> Map SS.LName SS.Schema
+primValueTypeEnv primsAvail = fmap primitiveType (filterAvailPrims primsAvail primitives)
+
+primNamedTypeEnv :: Set PrimitiveLifecycle -> Map SS.Name SS.NamedType
+primNamedTypeEnv primsAvail = fmap primtypeType (filterAvailTypes primsAvail primtypes)
 
 valueEnv :: Set PrimitiveLifecycle -> Options -> BuiltinContext -> Map SS.LName Value
-valueEnv primsAvail opts bic = fmap f (filterAvail primsAvail primitives)
+valueEnv primsAvail opts bic = fmap f (filterAvailPrims primsAvail primitives)
   where f p = (primitiveFn p) opts bic
```

Comment:
It strikes me that we now have two different functions (`primValueTypeEnv` and `valueEnv`) which separately call `filterAvailPrims primsAvail primitives` and then extract out the relevant things that they need (the `Schema`s and `Value`s, respectively). This feels a bit wasteful, since we really only need to perform the filtering once: we always call `primValueTypeEnv` and `valueEnv` in the same parts of the code, and always with the same `primsAvail`.

In light of this, what do you think of refactoring these functions to take in the filtered `Map`s explicitly? That is, something like:

```hs
primValueTypeEnv :: Map SS.LName Primitive -> Map SS.LName SS.Schema
valueEnv :: Options -> BuiltinContext -> Map SS.LName Primitive -> Map SS.LName Value
```

And then at call sites, the `Map SS.LName Primitive` argument would always be derived from `filterAvailPrims`, e.g.,

```hs
       let primsEnv = filterAvailPrims primsAvail primitives
       let rw0 = TopLevelRW
          { rwValues     = valueEnv opts bic primsEnv
          , ...
          , rwValueTypes = primValueTypeEnv primsEnv
          , ...
          }
```

We could consider performing a similar refactoring to `primNamedTypeEnv` as well for consistency, although this isn't as important, given that that is the only function that calls `filterAvailTypes`.